FW/FreeRTOS/Task.hpp

```diff
@@ -0,0 +1,146 @@
+#ifndef _FREERTOS_TASK_HPP
+#define _FREERTOS_TASK_HPP 
+
+#include "FreeRTOS.h"
+#include "Task.h"
+#include "string.h"
+
+namespace FreeRTOS
+{
+    /**
+     * @brief FreeRTOS task wrapper
+     * @note Template stackSize have to be a multiple of sizeof(StackType_t)
+     * 
+     * @tparam stackSize 
+     */
+    template <size_t stackSize>
+    class Task {
+    public:
+
+        /**
+         * @brief Construct a new Task:: Task object
+         * 
+         * @param name Task name, will be truncated if longer than configMAX_TASK_LENGTH
+         * @param stackSize Stack size in bytes, rounded up to nearest sizeof(portSTACK_TYPE)
+         * @param priority Task priority
+         */
+        Task(const char *name, UBaseType_t priority) {
+            // TODO: assert(stackSize % sizeof(StackType_t));
+
+            strncpy(m_name, name, configMAX_TASK_NAME_LEN);
+            // strncpy() does not guarantee null-termination.
+            m_name[configMAX_TASK_NAME_LEN - 1] = '\0';
+
+            // FreeRTOS takes stack size in words of length sizeof(portSTACK_TYPE)
+            // round up
+            m_stackSize = (configSTACK_DEPTH_TYPE)((stackSize + sizeof(portSTACK_TYPE) - 1) / sizeof(portSTACK_TYPE));
+
+            m_priority = priority;
+            m_periodInitialized = false;
+        }
+
+
+        /**
+         * @brief Destroy the Task object
+         * 
+         */
+        ~Task() {
+            DeInit();
+            vTaskDelete(m_handle);
+        }
+
+
+        /**
+         * @brief Start executing task.
+         * 
+         */
+        void Start()  {
+           m_handle = xTaskCreateStatic(Task::TaskFunction, m_name, sizeof(m_stack) / sizeof(StackType_t), this, m_priority, m_stack, &m_buffer);
+        }
+
+
+         /**
+         * @brief Start FreeRTOS scheduler. 
+         *        NOTE! This function shall never return under normal circumstances.
+         * 
+         */
+        static void StartScheduler() {
+            vTaskStartScheduler();
+        }
+
+
+        // Four prio-levels should be enough for now
+        static constexpr UBaseType_t IdlePriority   = tskIDLE_PRIORITY;
+        static constexpr UBaseType_t LowPriority    = tskIDLE_PRIORITY + 1;
+        static constexpr UBaseType_t MediumPriority = tskIDLE_PRIORITY + 2;
+        static constexpr UBaseType_t HighPriority   = tskIDLE_PRIORITY + 3;
+    protected:
+
+        virtual void Init() {};
+        virtual void Run() = 0;
+        virtual void DeInit(){};
+        
+        
+        /**
+         * @brief Initialize a periodic delay, useful for having a periodic execution in task
+         *        Use EndPeriod() to wait for next period
+         * 
+         * @param milliseconds 
+         */
+        void InitPeriod(uint32_t milliseconds)  {
```

Comment:
I'd suggest StartPeriod(), as init sounds like the function will set things up in preparation for use, rather than actually start the period. Start and end are good antonyms.

Just to understand, is this used to have a minimum time between periodic events, e.g. to make sure you don't invoke a method more often than, say once every 5 seconds?

Reply:
One of the more common constructs I've seen in RTOS-work is to have an endless loop running at a fixed period independent on how much time the execution takes inside the loop. By having a simple delay, the period will be execution time + delay, which could be non-deterministic. This is a helper to realize such a loop. `InitPeriod(time)` before the loop, and then end the loop with `EndPeriod()` to delay `time - execution time`.

Reply:
OK, I understand. I've done similar things with a timeout calculated in advance, doing some work and then waiting for events until the pre-calculated timeout hits. I just wasn't familiar with the start/end period way of looking at it.